### PR TITLE
Aes cbc pad

### PR DESCRIPTION
**Pull request by williamcroberts**

### CODE REVIEW

File: src/lib/tpm.c [hunk 1]
```diff
@@ -2032,8 +2044,33 @@ CK_RV tpm_decrypt(crypto_op_data *opdata,
     twist auth = tpm_enc_data->tobj->unsealed_auth;
     ESYS_TR handle = tpm_enc_data->tobj->tpm_handle;
 
-    return encrypt_decrypt(ctx, handle, auth, mode, DECRYPT,
-            iv, ctext, ctextlen, ptext, ptextlen);
+    /*
+     * if we're stripping padding use a non-user supplied buffer so they don't
+     * need to allocae one bigenough to include the padding.
```
Comment (fishilico) [inline src/lib/tpm.c:2049]: Misspelling: `allocae one bigenough` -> `allocate one big enough`

File: src/lib/utils.c [hunk 2]
```diff
@@ -454,3 +454,46 @@ CK_RV ec_params_to_nid(CK_ATTRIBUTE_PTR ecparams, int *nid) {
...
+        CK_BYTE_PTR out, CK_ULONG_PTR outlen) {
+
+    if (inlen % 16) {
+        LOGE("AES_CBC_PAD data should be block sized, got: %lu", inlen);
+        return CKR_ENCRYPTED_DATA_LEN_RANGE;
+    }
+
+    CK_BYTE pad_value = in[inlen - 1];
+    if (pad_value < 1 || pad_value > 16) {
+        LOGE("Nonesensical pad value, got: %u, expected 1-16", pad_value);
```
Comment (fishilico) [inline src/lib/utils.c:485]: Misspelling: `Nonesensical` -> `Nonsensical`

File: src/lib/utils.c [hunk 3]
```diff
@@ -454,3 +454,46 @@ CK_RV ec_params_to_nid(CK_ATTRIBUTE_PTR ecparams, int *nid) {
...
+    memset(&out[inlen], pad_value, pad_value);
+    *outlen = total_bytes;
+
+    return CKR_OK;
+}
+
+CK_RV remove_pkcs7_pad(CK_BYTE_PTR in, CK_ULONG inlen,
+        CK_BYTE_PTR out, CK_ULONG_PTR outlen) {
+
+    if (inlen % 16) {
```
Comment (fishilico) [inline src/lib/utils.c:478]: Maybe also verify that `inlen != 0` (or that `inlen >= 16`)?

File: src/lib/utils.h [hunk 4]
```diff
@@ -84,6 +84,36 @@ CK_RV utils_ctx_wrap_objauth(twist wrappingkey, twist objauth, twist *wrapped_au
  */
 CK_RV ec_params_to_nid(CK_ATTRIBUTE_PTR ecparams, int *nid);
 
+/**
+ * Applies a PKCS5 padding to a blocksize of 16.
```
Comment (fishilico) [inline src/lib/utils.h:88]: It is strange to document "PKCS5" while the function is named "PKCS7". Maybe you could make this more uniform?

**Review by fishilico — COMMENTED** [review]
Hi, I found some minor things to say about this PR. Nothing really important, but this could help improving it.

**Comment by williamcroberts** [comment]
Marked this as a draft, its not even close to ready for prime time.
1. Callers can call into AES with buffers larger than TPM2B_MAX which will cause a failure. 
2. C_EncryptFinal doesn't handle if the buffer is block size units, it needs to return a full block of padding.

**Comment by williamcroberts** [comment]
Not even close to ready, but I have picked up this work on the same branch on my fork aes-cbc-pad. Hoping to be able to merge this and CTR by end of next week.

**Pull request closed without merging**